netaddr.go

```diff
@@ -781,28 +780,28 @@ func FromStdAddr(stdIP net.IP, port int, zone string) (_ IPPort, ok bool) {
 	return IPPort{IP: ip, Port: uint16(port)}, true
 }
 
-var udpAddrPool = &sync.Pool{
-	New: func() interface{} { return new(net.UDPAddr) },
-}
-
 // UDPAddr returns a standard library net.UDPAddr from p.
 // The returned value is always non-nil. If p.IP is the zero
 // value, then UDPAddr.IP is nil.
 //
-// UDPAddr necessarily does two allocations. If you call PutUDPAddr
-// after you're done with it, though, then subsequent UDPAddr calls
-// can reuse the memory.
+// UDPAddr necessarily does two allocations. If you have an existing
+// UDPAddr already allocated, see UDPAddrReuse.
 func (p IPPort) UDPAddr() *net.UDPAddr {
-	ua := udpAddrPool.Get().(*net.UDPAddr)
-	ua.Port = int(p.Port)
-	ua.IP, ua.Zone = p.IP.ipZone(ua.IP)
-	return ua
+	ret := &net.UDPAddr{
+		Port: int(p.Port),
+	}
+	ret.IP, ret.Zone = p.IP.ipZone(nil)
+	return ret
 }
 
-// PutUDPAddr adds ua to an internal pool for later reuse by IPPort.UDPAddr.
-// Use of PutUDPAddr is optional; improper use can cause mysterious errors.
-// You must only call PutUDPAddr if there are no remaining references to ua.
-func PutUDPAddr(ua *net.UDPAddr) { udpAddrPool.Put(ua) }
+// UDPAddrReuse is like UDPAddr, but reuses the provided UDPAddr,
+// which must be non-nil. If reuse.IP has a capacity of 16,
+// UDPAddrReuse is allocation-free.
+func (p IPPort) UDPAddrReuse(reuse *net.UDPAddr) *net.UDPAddr {
```

Comment:
do we even need to return it? the caller has it. But then it needs a new name which probably doesn't sort next to UDPAddr. Hm, `UDPAddrPopulate`?

In any case, I much prefer having a separate method for the weird memory-reuse case rather than unifying them into one and forcing normal people to pass mystery `nil` values.

Reply:
Returning it is nice for compactness when sending, the caller can `sock.WriteToUDP(bs, ipp.UDPAddr(reuse))`. I'm also fine with not returning if you find that too weird, but I think it's nice to make the method pass-through.

UDPAddrPopulate or UDPAddrFill sgtm if we don't return.

Reply:
Yeah, it'll probably be more convenient. I'd just thought that the caller was going to need to do some sync.Pool work anyway, but it'd probably be like:

```go
    ua := pool.Get().(*net.UDPAddr)
    defer pool.Put(ua)
    sock.WriteToUDP(s, ipp.UDPAddrAt(ua))
```

And I'm thinking I like the name `UDPAddrAt`, like https://golang.org/pkg/reflect/#NewAt (like C++ "[placement new](https://en.cppreference.com/w/cpp/language/new#Placement_new)")

Reply:
UDPAddrAt sgtm, renamed.